File: mt5-core-server/src/services/auto/analyzers/smc/liquidityZones.ts
/**
 * SMC Engine V2 — Liquidity Zones Detection
 * Port จาก Pine Script V8.3: Equal H/L, Swing, MTF merge, Confluence Stars
 * 2026-05-06
 */

import type { Candle, LiquidityZone, OrderBlock, MarketStructure, PremiumDiscountZone } from './types.js';
import { detectSwings, computeATR } from './utils.js';

export interface LiquidityConfig {
  lookback: number;
  equalThresholdPct: number;   // % threshold for equal H/L (default 0.1%)
  swingLength: number;
  maxZones: number;
}

const DEFAULT_CFG: LiquidityConfig = { lookback: 10, equalThresholdPct: 0.1, swingLength: 10, maxZones: 50 };

export function detectLiquidityZones(
  candles: Candle[],
  config: Partial<LiquidityConfig> = {}
): LiquidityZone[] {
  const cfg = { ...DEFAULT_CFG, ...config };
  if (candles.length < 20) return [];
  const zones: LiquidityZone[] = [];

  // 1. Equal Highs
  for (let i = 1; i <= Math.min(cfg.lookback, candles.length - 1); i++) {
    const ch = candles[candles.length - 1 - i]?.h;
    if (!ch) continue;
    const thr = ch * (cfg.equalThresholdPct / 100);
    let eqCount = 0;
    for (let j = i + 1; j <= Math.min(i + cfg.lookback, candles.length - 1); j++) {
      const other = candles[candles.length - 1 - j]?.h;
      if (other && Math.abs(other - ch) <= thr) eqCount++;
    }
    if (eqCount >= 1 && !zones.some(z => z.isHigh && Math.abs(z.price - ch) <= thr)) {
      zones.push({ price: ch, barIndex: candles.length - 1 - i, isHigh: true, swept: false, strength: eqCount, source: 'EQUAL_HL', confluenceStars: 0 });
    }
  }

  // 2. Equal Lows
  for (let i = 1; i <= Math.min(cfg.lookback, candles.length - 1); i++) {
    const cl = candles[candles.length - 1 - i]?.l;
    if (!cl) continue;
    const thr = cl * (cfg.equalThresholdPct / 100);
    let eqCount = 0;
    for (let j = i + 1; j <= Math.min(i + cfg.lookback, candles.length - 1); j++) {
      const other = candles[candles.length - 1 - j]?.l;
      if (other && Math.abs(other - cl) <= thr) eqCount++;
    }
    if (eqCount >= 1 && !zones.some(z => !z.isHigh && Math.abs(z.price - cl) <= thr)) {
      zones.push({ price: cl, barIndex: candles.length - 1 - i, isHigh: false, swept: false, strength: eqCount, source: 'EQUAL_HL', confluenceStars: 0 });
    }
  }

  // 3. Swing H/L Liquidity
  const { highs, lows } = detectSwings(candles, cfg.swingLength);
  for (const sh of highs.slice(-5)) {
    if (!zones.some(z => z.isHigh && Math.abs(z.price - sh.price) < sh.price * 0.001)) {
      zones.push({ price: sh.price, barIndex: sh.index, isHigh: true, swept: false, strength: 2, source: 'SWING', confluenceStars: 0 });
    }
  }
  for (const sl of lows.slice(-5)) {
    if (!zones.some(z => !z.isHigh && Math.abs(z.price - sl.price) < sl.price * 0.001)) {
      zones.push({ price: sl.price, barIndex: sl.index, isHigh: false, swept: false, strength: 2, source: 'SWING', confluenceStars: 0 });
    }
  }

  // 4. Mark swept zones
  const last = candles[candles.length - 1];
  for (const z of zones) {
    if (z.isHigh && last.h > z.price) z.swept = true;
    if (!z.isHigh && last.l < z.price) z.swept = true;
  }

  return zones.filter(z => !z.swept).slice(0, cfg.maxZones);
}

/**
 * Score zone confluence (0-5 stars)
 * Based on Pine Script V8.3 confluence scoring
 */
export function scoreZoneConfluence(
  zone: LiquidityZone,
  bullOBs: OrderBlock[],
  bearOBs: OrderBlock[],
  structure: MarketStructure,
  pdZone: PremiumDiscountZone,
  atr: number
): number {
  let score = 1; // Base score

  // OB overlap (+2)
  const obs = zone.isHigh ? bearOBs : bullOBs;
  if (obs.some(ob => zone.price <= ob.top * 1.001 && zone.price >= ob.bottom * 0.999)) score += 2;

  // Structure confluence (+1)
  if (Math.abs(zone.price - structure.structureHigh) < atr * 0.5 ||
      Math.abs(zone.price - structure.structureLow) < atr * 0.5) score += 1;

  // Premium/Discount alignment (+1)
  if (zone.isHigh && pdZone === 'PREMIUM') score += 1;
  if (!zone.isHigh && pdZone === 'DISCOUNT') score += 1;

  // Trend alignment (+1)
  if ((zone.isHigh && structure.direction === 'BEARISH') ||
      (!zone.isHigh && structure.direction === 'BULLISH')) score += 1;

  return Math.min(5, score);
}

File: composeApp/src/commonMain/kotlin/com/example/personalaibot/automation/smc/LiquidityZones.kt
package com.example.personalaibot.automation.smc

import com.example.personalaibot.tools.trading.Candle
import kotlin.math.abs
import kotlin.math.min

/**
 * Liquidity Zones — port จาก mt5-core-server liquidityZones.ts (Pine V8.3)
 * Equal Highs/Lows + Swing liquidity + confluence stars (0-5★)
 */
internal object LiquidityZones {

    fun detect(
        candles: List<Candle>,
        lookback: Int = 10,
        equalThresholdPct: Double = 0.1,
        swingLength: Int = 10,
        maxZones: Int = 50
    ): List<LiquidityZone> {
        if (candles.size < 20) return emptyList()
        val zones = mutableListOf<LiquidityZone>()
        val n = candles.size

        // 1. Equal Highs
        for (i in 1..min(lookback, n - 1)) {
            val ch = candles[n - 1 - i].high
            val thr = ch * (equalThresholdPct / 100.0)
            var eqCount = 0
            for (j in i + 1..min(i + lookback, n - 1)) {
                val other = candles[n - 1 - j].high
                if (abs(other - ch) <= thr) eqCount++
            }
            if (eqCount >= 1 && zones.none { it.isHigh && abs(it.price - ch) <= thr }) {
                zones += LiquidityZone(ch, n - 1 - i, isHigh = true, strength = eqCount, source = "EQUAL_HL")
            }
        }

        // 2. Equal Lows
        for (i in 1..min(lookback, n - 1)) {
            val cl = candles[n - 1 - i].low
            val thr = cl * (equalThresholdPct / 100.0)
            var eqCount = 0
            for (j in i + 1..min(i + lookback, n - 1)) {
                val other = candles[n - 1 - j].low
                if (abs(other - cl) <= thr) eqCount++
            }
            if (eqCount >= 1 && zones.none { !it.isHigh && abs(it.price - cl) <= thr }) {
                zones += LiquidityZone(cl, n - 1 - i, isHigh = false, strength = eqCount, source = "EQUAL_HL")
            }
        }

        // 3. Swing H/L liquidity (5 จุดล่าสุด)
        val (highs, lows) = detectSwings(candles, swingLength)
        for (sh in highs.takeLast(5)) {
            if (zones.none { it.isHigh && abs(it.price - sh.price) < sh.price * 0.001 }) {
                zones += LiquidityZone(sh.price, sh.index, isHigh = true, strength = 2, source = "SWING")
            }
        }
        for (sl in lows.takeLast(5)) {
            if (zones.none { !it.isHigh && abs(it.price - sl.price) < sl.price * 0.001 }) {
                zones += LiquidityZone(sl.price, sl.index, isHigh = false, strength = 2, source = "SWING")
            }
        }

        // 4. mark swept ด้วยแท่งล่าสุด แล้วตัดทิ้ง
        val last = candles[n - 1]
        for (z in zones) {
            if (z.isHigh && last.high > z.price) z.swept = true
            if (!z.isHigh && last.low < z.price) z.swept = true
        }

        return zones.filter { !it.swept }.take(maxZones)
    }

    /** Confluence score 0-5★ — ตรงต้นฉบับ: base 1 + OB overlap(+2) + structure(+1) + PD(+1) + trend(+1) */
    fun scoreConfluence(
        zone: LiquidityZone,
        bullObs: List<SmcOrderBlock>,
        bearObs: List<SmcOrderBlock>,
        structure: MarketStructure,
        pdZone: PremiumDiscountZone,
        atr: Double
    ): Int {
        var score = 1

        val obs = if (zone.isHigh) bearObs else bullObs
        if (obs.any { zone.price <= it.top * 1.001 && zone.price >= it.bottom * 0.999 }) score += 2

        if (abs(zone.price - structure.structureHigh) < atr * 0.5 ||
            abs(zone.price - structure.structureLow) < atr * 0.5
        ) score += 1

        if (zone.isHigh && pdZone == PremiumDiscountZone.PREMIUM) score += 1
        if (!zone.isHigh && pdZone == PremiumDiscountZone.DISCOUNT) score += 1

        if ((zone.isHigh && structure.direction == StructureDirection.BEARISH) ||
            (!zone.isHigh && structure.direction == StructureDirection.BULLISH)
        ) score += 1

        return min(5, score)
    }
}
